### ISSUE TITLE
Report what connection spec failed on `ActiveRecord::StatementInvalid`

### ISSUE THREAD

**Issue by sobrinho** [rails/rails#49331]
### Steps to reproduce
Backtraces doesn't always point exactly to what connection have failed.

```ruby
ActiveRecord::Base.configurations.configs_for(env_name: ActiveRecord::Tasks::DatabaseTasks.env).each do |db_config|
  ActiveRecord::Base.establish_connection(db_config)
  puts ActiveRecord::Base.connection.select_values 'select now()'
end
```

### Expected behavior
```
ActiveRecord::StatementInvalid
primary --> [connection name or something about what connection failed]
PG::ConnectionBad: PQconsumeInput() FATAL:  terminating connection due to idle-in-transaction timeout
ERROR:  server conn crashed?
SSL connection has been closed unexpectedly
```

### Actual behavior
```
ActiveRecord::StatementInvalid
PG::ConnectionBad: PQconsumeInput() FATAL:  terminating connection due to idle-in-transaction timeout
ERROR:  server conn crashed?
SSL connection has been closed unexpectedly
PG::ConnectionBad
```

### System configuration
**Rails version**: 6.1.7.5

**Ruby version**: 2.7.6

**Comment by fatkodima** [comment]
Agreed on the usefulness. Took a stab at it - https://github.com/rails/rails/pull/49337.

**Comment by eileencodes** [comment]
What's the use case to know the connection where the query failed? Applications don't usually establish a connection right before running a query, that's the job of the connection model to do. When you have a StatementInvalid error it's usually bad syntax or something else to do with the query rather than the connection. I don't feel like a lot of value is added in adding this information since it doesn't help you figure out what's wrong with the query.

**Comment by fatkodima** [comment]
A few examples I can think of (not sure how realistic):

1. An example from the linked issue (https://github.com/rails/rails/issues/49331) - if we have configured timeouts on some dbs, but forgot on others, then the thrown error will point to the right db and speed up debugging (investigation)
2. If we have similarly named tables in a few dbs, then the error thrown (for example, for mismatched foreign keys or not null violation) will point to the right database
3. If we have migration ran on a primary and for some reason it was not propagated to the replica and is not detected, then we will quickly get the info about the erroring db 

If we have only 1 database, that is definitely not useful.

**Comment by eileencodes** [comment]
Oh I just realized we already have the connection pool set on the error in 7.1 when https://github.com/rails/rails/pull/48295 was implemented.

**Comment by matthewd** [comment]
Yeah, I think I'm still at https://github.com/rails/rails/pull/48295#issuecomment-1570721170

Inserting it into the message isn't totally unreasonable (though I think putting it before the actual error, as #49337 currently does, _is_ unreasonable)... but I'd really prefer to explore non-message-mangling ways we can convey additional exception-carried context through error reporting.

---

For example, what would it look like if we invented a hash-returning `Exception#context` method, and then taught our error reporting to include that information?

I was originally thinking this would exist _next to_ ErrorReporter's context (meaning this method should have a different name), and I think I'm still slightly leaning that way (keeping "extra information about what went wrong" separate from "extra information about what we were doing at the time"), but it also seems possible they're the same thing, and ErrorReporter could just:

```
full_context = full_context.merge(error.context) if error.respond_to?(:context)
```

cc @byroot 🤔

**Comment by byroot** [comment]
> 🤔

So, IMO including the pool name in AR error messages make sense, as in multi-db scenarios, it allows to see at a glance if the errors are localized to a specific pool.

That said, I assume 99%+ of Rails app only have one database.

> what would it look like if we invented a hash-returning Exception#context method, and then taught our error reporting to include that information?

Yeah, I like that. Maybe something a bit less generic than `#context` to avoid conflicts, but it's a very nice interface to extend error reports.

**Comment by sobrinho** [comment]
This is all about being quickly when errors happens.

For instance, if I need to check the backtrace, at least I have to open the exception in my exception tracker platform.

If it's right on the exception message, it can be read right away from chat tools, email subject and etc, no additional steps.

Also, if not using the exception message and instead we need to capture that info from the exception and propagate somehow to the exception tracker mechanism:

1. We need to do that in different ways for different exception platforms
2. The same will happen for different projects
3. And it also might change from version to version on Rails upgrades

All of that doesn't happen if the exception message is changed from:

```
ActiveRecord::StatementInvalid
PG::ConnectionBad: PQconsumeInput() FATAL:  terminating connection due to idle-in-transaction timeout
ERROR:  server conn crashed?
SSL connection has been closed unexpectedly
PG::ConnectionBad
```

To (or similar):

```
ActiveRecord::StatementInvalid
SPEC: primary
PG::ConnectionBad: PQconsumeInput() FATAL:  terminating connection due to idle-in-transaction timeout
ERROR:  server conn crashed?
SSL connection has been closed unexpectedly
```

**Comment by sobrinho** [comment]
Our use case:

```ruby
ActiveRecord::Base.transaction do # primary
  Auditing.transaction do # another one
    # PG::ConnectionBad: PQconsumeInput() FATAL:  terminating connection due to idle-in-transaction timeout
    # which one did raise that?
  end
end
```

**Comment by byroot** [comment]
> This is all about being quickly when errors happens.

Yes I agree. If you see a spike of error, it's nice to be able to tell at a glance which DB is down.

> Our use case:

So it's totally orthogonal to the issue you filed, but note that this is an anti-pattern, you are nesting two transaction from two systems, which means if one deadlock it can also bring down the other one with it.

**Comment by sobrinho** [comment]
That's the idea indeed, it's not perfect but I want the primary transaction to fail if the secondary transaction fails for whatever reason.

There's a small possibility of a race condition that the second (or inner transaction, depending on how you read it) can complete but not the first one but it's very rare and never happened to us.

**Comment by byroot** [comment]
> There's a small possibility of a race condition that the second

I'm not talking of race conditions, I'm talking of holding locks on two distinct servers. This works fine most of the time but can lead to catastrophic outages. But that's not the topic of this issue, so I won't insist.
